### PR TITLE
 add restore volume from snapshot for hostpath

### PR DESCRIPTION
**Pull request by wackxu**
based on https://github.com/kubernetes-csi/drivers/pull/94

### CODE REVIEW

File: pkg/hostpath/controllerserver.go [hunk 1]
```diff
@@ -84,6 +90,23 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		glog.V(3).Infof("failed to create volume: %v", err)
 		return nil, err
 	}
+	if req.GetVolumeContentSource() != nil {
+		contentSource := req.GetVolumeContentSource()
+		if contentSource.GetSnapshot() != nil {
+			snapshotId := contentSource.GetSnapshot().GetId()
+			snapshot, ok := hostPathVolumeSnapshots[snapshotId]
```
Comment (lpabon) [inline pkg/hostpath/controllerserver.go]: We keep them in memory? So if we restart the driver they get lost, right? if this is is true, we should document this somewhere. I am ok with having it memory and documenting that this is a test-only driver.

File: pkg/hostpath/controllerserver.go [hunk 2]
```diff
@@ -84,6 +90,23 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+			if !ok {
+				return nil, status.Errorf(codes.NotFound, "cannot find snapshot %v", snapshotId)
+			}
+			snapshotPath := snapshot.Path
+			args := []string{"zxvf", snapshotPath, "-C", path}
```
Comment (lpabon) [inline pkg/hostpath/controllerserver.go]: nice snapshot method 👍

File: pkg/hostpath/controllerserver.go [hunk 3]
```diff
@@ -139,3 +162,227 @@ func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req
...
+	}
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid delete snapshot req: %v", req)
+		return nil, err
+	}
+	snapshotID := req.GetSnapshotId()
+	glog.V(4).Infof("deleting volume %s", snapshotID)
+	path := snapshotRoot + snapshotID + ".tgz"
+	os.RemoveAll(path)
```
Comment (lpabon) [inline pkg/hostpath/controllerserver.go]: Could log if the file not file for some reason. May be a good idea to return error if not found, which would be good for e2e tests.

**Comment by lpabon** [comment]
Could you also add an issue to add unit tests to hostpath driver? We need to start adding tests.

File: pkg/hostpath/controllerserver.go [hunk 4]
```diff
@@ -84,6 +90,23 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
 	}
+	if req.GetVolumeContentSource() != nil {
+		contentSource := req.GetVolumeContentSource()
+		if contentSource.GetSnapshot() != nil {
+			snapshotId := contentSource.GetSnapshot().GetId()
+			snapshot, ok := hostPathVolumeSnapshots[snapshotId]
+			if !ok {
+				return nil, status.Errorf(codes.NotFound, "cannot find snapshot %v", snapshotId)
+			}
+			snapshotPath := snapshot.Path
```
Comment (sarjeet2013) [inline pkg/hostpath/controllerserver.go]: Do we also check the snapshot status before using it? This won't matter here, but a good practice if this example is being picked for other places.

**Comment by sarjeet2013** [comment]
LGTM

File: pkg/hostpath/controllerserver.go (hunk 4 above)
Comment (wackxu) [reply pkg/hostpath/controllerserver.go]: Thanks for your review, updated

**Pull request merged**